### PR TITLE
Support 2-phase transaction apply in staged ledger

### PR DESCRIPTION
**Pull request by nholland94**
Apply zkapp transactions in two phases. Apply all the simple commands and fee payer part of zkapp transactions and in the second phase apply the zkapp account updates

There's a couple of stubs in non-staged-ledger code, and a few small TODOs to fill in (many of which aren't actually important for testing)

### CODE REVIEW

File: src/lib/staged_ledger/staged_ledger.ml [hunk 1]
```diff
@@ -11,11 +11,43 @@ open Signature_lib
 module Ledger = Mina_ledger.Ledger
 module Sparse_ledger = Mina_ledger.Sparse_ledger
 
+(* TODO: move me *)
+let result_list_fold ls ~init ~f =
```
Comment (psteckler) [inline src/lib/staged_ledger/staged_ledger.ml:15]: these could be in the new `Mina_stdlib`

File: src/lib/transaction_logic/mina_transaction_logic.ml [hunk 2]
```diff
@@ -6,6 +6,23 @@ open Mina_transaction
 module Zkapp_command_logic = Zkapp_command_logic
 module Global_slot = Mina_numbers.Global_slot
 
+(* TODO: move me *)
```
Comment (psteckler) [inline src/lib/transaction_logic/mina_transaction_logic.ml:9]: `Mina_stdlib`

File: src/lib/transaction_logic/mina_transaction_logic.ml [hunk 3]
```diff
@@ -338,6 +357,41 @@ module type S = sig
       }
   end
 
+  module Transaction_partially_applied : sig
+    module Zkapp_command_partially_applied : sig
```
Comment (psteckler) [inline src/lib/transaction_logic/mina_transaction_logic.ml:361]: maybe `fee_payer` instead of `partially` ?

**Review by psteckler — COMMENTED** [review]
Am I right that `phase_1` and `first_pass` are almost-synonymous? Can we use a single term to describe that (ditto for 2/second).

File: src/lib/crypto/kimchi_bindings/stubs/kimchi_types.ml [hunk 4]
```diff
@@ -206,11 +193,9 @@ module VerifierIndex = struct
       ; lookup_info : lookup_info
       ; runtime_tables_selector : 'poly_comm option
       }
-    [@@boxed]
   end
 
   type nonrec 'fr domain = { log_size_of_group : int; group_gen : 'fr }
-  [@@boxed]
```
Comment (ghost-not-in-the-shell) [inline src/lib/crypto/kimchi_bindings/stubs/kimchi_types.ml:199]: Those changes seem like a side effect of running unit tests. I also saw those changes being made after I ran `testone.sh` script.

Comment (ghost-not-in-the-shell) [reply src/lib/crypto/kimchi_bindings/stubs/kimchi_types.ml:199]: I mean remove the `[@@boxed]`

File: src/lib/transaction_snark/test/util.ml [hunk 5]
```diff
@@ -436,6 +441,9 @@ let check_balance pk balance ledger =
   [%test_eq: Balance.t] acc.balance (Balance.of_nanomina_int_exn balance)
 
 (** Test legacy transactions*)
+let test_transaction_union ?expected_failure:_ ?txn_global_slot:_ _ _ =
+  failwith "TODO"
```
Comment (ghost-not-in-the-shell) [inline src/lib/transaction_snark/test/util.ml:445]: Should we just delete this?

**Pull request merged**